src/qililab/qprogram/qprogram.py

```diff
@@ -101,6 +102,54 @@ def __init__(self) -> None:
         self._variables: list[Variable] = []
         self._block_stack: deque[Block] = deque([self._body])
 
+    def __dict__(self) -> dict:  # type: ignore[override]
+        """Dictionary representation"""
+        return self._to_dict(self)
+
+    def _to_dict(self, program) -> dict:
+        """Converts qprogram to dictionary containing the instructions
+
+        Args:
+            program (QProgram): qprogram to convert
+
+        Returns:
+            dict: qprogram dictionary
+        """
+        if hasattr(program, "body"):
+            return self._to_dict(program.body)
+
+        temp_dict = {"name": type(program).__name__}
+        for field in fields(program):
+            if field.name in ["_uuid", "variable"]:
+                continue
+            temp_dict[field.name] = (
+                getattr(program, field.name)
+                if field.name != "elements"
+                else [self._to_dict(element) for element in getattr(program, field.name)]
+            )
+
+        if isinstance(program, Operation):
+            if hasattr(program, "waveform"):
+                temp_dict["waveform"] = (
+                    [(type(program.waveform).__name__, program.waveform.envelope())]
+                    if isinstance(program.waveform, Waveform)
+                    else [
+                        (type(program.waveform.I).__name__, program.waveform.I.envelope()),  # type: ignore[assignment]
+                        (type(program.waveform.Q).__name__, program.waveform.Q.envelope()),  # type: ignore[assignment]
+                    ]
+                )
+            if hasattr(program, "weights"):
+                temp_dict["weights"] = [  # type: ignore[assignment]
+                    (type(program.weights.I).__name__, program.weights.I.envelope()),
+                    (type(program.weights.Q).__name__, program.weights.Q.envelope()),
+                ]
+
+        for key, value in temp_dict.copy().items():
+            if "UUID" in str(value):
+                temp_dict[key] = None
+
+        return temp_dict
```

Comment:
Also, why do we remove several fields like `UUID`? `__dict__` should return the dictionary representation of the class as it is. In other words we could create a `from_dict()` method that takes this dictionary and correctly builds the class instance again. This will not be the case with this implementation.

Reply:
I also commented above. If we need the strings we should implement the `__str__` method, which is ok to hide information and display only the useful ones. `__dict__` by python documentation should return all attributes of the class.